go/pkg/client/tree.go

```diff
@@ -65,17 +81,22 @@ func shouldIgnore(inp string, t command.InputType, excl []*command.InputExclusio
 
 // loadFiles reads all files specified by the given InputSpec (descending into subdirectories
 // recursively), and loads their contents into the provided map.
-func loadFiles(execRoot string, excl []*command.InputExclusion, path string, fs map[string]*fileSysNode, chunkSize int, cache filemetadata.Cache) error {
+func loadFiles(execRoot string, excl []*command.InputExclusion, path string, fs map[string]*fileSysNode, chunkSize int, cache filemetadata.Cache, opts TreeSymlinkOpts) error {
 	absPath := filepath.Clean(filepath.Join(execRoot, path))
 	meta := cache.Get(absPath)
 	t := command.FileInputType
-	if smd := meta.Symlink; smd != nil && smd.IsDangling {
-		return nil
-	}
-	if meta.Err != nil {
+	isSymlink := meta.Symlink != nil
+	isDanglingSymlink := isSymlink && meta.Symlink.IsDangling
+	if isDanglingSymlink {
+		if !(opts.Preserved && opts.AllowDanglingSymlink) {
+			return nil
```

Comment:
symlink is ignored silently here?

Reply:
Yeah, i think the behavior is consistent with the old code:

* if we preserved symlink, and allow for dangling symlinks, then the produced tree will contain a `SymlinkNode`, which points to an invalid target.
* Otherwise, return `nil`, which is what the old code did.

Reply:
Then I'm wondering what value does `AllowDanglingSymlink` have. Also `IgnoreDanglingSymlink` sounds proper name?

Reply:
Yeah. Given how `loadFile` is implemented, if the user doesn't specify the target, a symlink effectively becomes "dangling" in the tree. I removed this option. It seems that neither the old code nor the new one needs to treat dangling symlinks as an error.

Reply:
update PR description then?

Reply:
Done

Reply:
I like it, but for the record, I just realized (a tad too late), that we could merge the upcoming compression change and this change without breaking compatibility using your original idea. I want to ask you if you guys still prefer it, because we could revert #235 and do it:

If we introduce TreeConstructionOption similar to client.Opt (or as you did), then we can have our new CompressionCutoff as part of these options and add these options to ConstructMerkleTree without breaking compatibility and without the need to move the tree package into client.

The cons of that option: it will result in a bit of a duplicated code for us when creating the compressed Chunker; it necessitates more work now (reverting #235 and what you did here, again); the client Options will be broken into two packages/types with similar code and the caller will need to *translate* from one to another (because ultimately, with all these options, they are Client options set through flags).

The pros of this option:
- won't break compatibility
- it passes the minimal required thing (only the construction options, not the whole client), allowing us to reuse the tree functions in a server implementation more easily (e.g. fake server package)

I'm not sure what we should do. I'm okay with both. What do you guys think?

Reply:
> it will result in a bit of a duplicated code for us when creating the compressed Chunker

I think this will be minimal and worth the effort (a dozen lines of dupped at most?)

I don't like the options translation as much but I generally like having tree be independent of Client - albeit I can't really see independent use cases, they're pretty independent in concept. Moreover, I *do* like not breaking compatibility. But I'm also fine with leaving #235 as is.

Reply:
I already manged to update deps for the PR in our repo, so I don't need the revert of the PR.

Reply:
> I want to ask you if you guys still prefer it, because we could revert #235 and do it:

Like @atetubou has said, since we've already resolved the incompatibility issue, I think it's fine to go with this new code, unless there are other concerns here?

Reply:
Great! Thank you, just checking!